File: deserialize.py
def deserialize(json_str):
    # Strip leading and trailing whitespace
    json_str = json_str.strip()
    
    # Handle null
    if json_str == 'null':
        return None
    
    # Handle booleans
    elif json_str == 'true':
        return True
    elif json_str == 'false':
        return False
    
    # Handle numbers (integers or floats)
    elif json_str.isdigit() or (json_str[0] == '-' and json_str[1:].isdigit()):
        return int(json_str)  # Handle integers
    
    elif '.' in json_str and json_str.replace('.', '', 1).isdigit():
        return float(json_str)  # Handle floats
    
    # Handle strings
    elif json_str.startswith('"') and json_str.endswith('"'):
        return json_str[1:-1]  # Remove surrounding quotes
    
    # Handle objects (dictionaries)
    elif json_str.startswith('{') and json_str.endswith('}'):
        obj = {}
        inner_str = json_str[1:-1].strip()  # Remove the curly braces
        key_value_pairs = inner_str.split(',')
        
        for pair in key_value_pairs:
            key, value = pair.split(":")
            obj[deserialize(key.strip())] = deserialize(value.strip())
        
        return obj
    
    # Handle arrays (lists)
    elif json_str.startswith('[') and json_str.endswith(']'):
        arr = []
        inner_str = json_str[1:-1].strip()  # Remove square brackets
        elements = inner_str.split(',')
        
        for element in elements:
            arr.append(deserialize(element.strip()))
        
        return arr
    
    else:
        raise ValueError(f"Invalid JSON string: {json_str}")

File: deserialize.go
package main

import (
	"errors"
	"strconv"
	"strings"
)

// Deserialize parses a JSON string into a Go value.
func Deserialize(json string) (interface{}, error) {
	json = strings.TrimSpace(json)

	// Handle string values
	if strings.HasPrefix(json, `"`) && strings.HasSuffix(json, `"`) {
		return json[1 : len(json)-1], nil
	}

	// Handle numeric values
	if num, _ := strconv.Atoi(json); num != 0 {
		return num, nil
	}
	if num, _ := strconv.ParseFloat(json, 64); num != 0 {
		return num, nil
	}

	// Handle boolean values
	switch json {
	case "true":
		return true, nil
	case "false":
		return false, nil
	}

	// Handle null value
	if json == "null" {
		return nil, nil
	}

	// Handle arrays
	if strings.HasPrefix(json, "[") && strings.HasSuffix(json, "]") {
		inner := json[1 : len(json)-1]
		elements := splitJSON(inner)
		var array []interface{}

		for _, element := range elements {
			val, _ := Deserialize(element)
			array = append(array, val)
		}

		return array, nil
	}

	// Handle objects
	if strings.HasPrefix(json, "{") && strings.HasSuffix(json, "}") {
		inner := json[1 : len(json)-1]
		pairs := splitJSON(inner)
		obj := make(map[string]interface{})

		for _, pair := range pairs {
			keyValue := strings.SplitN(pair, ":", 2)
			if len(keyValue) != 2 {
				return nil, errors.New("invalid JSON object")
			}

			key, _ := Deserialize(keyValue[0])
			value, _ := Deserialize(keyValue[1])

			strKey, ok := key.(string)
			if !ok {
				return nil, errors.New("invalid object key, must be a string")
			}

			obj[strKey] = value
		}

		return obj, nil
	}

	// Return error for invalid JSON
	return nil, errors.New("invalid JSON string")
}

// splitJSON splits JSON elements while accounting for nested structures.
func splitJSON(json string) []string {
	var elements []string
	start, braces, brackets := 0, 0, 0

	for i, char := range json {
		switch char {
		case '{':
			braces++
		case '}':
			braces--
		case '[':
			brackets++
		case ']':
			brackets--
		case ',':
			if braces == 0 && brackets == 0 {
				elements = append(elements, strings.TrimSpace(json[start:i]))
				start = i + 1
			}
		}
	}

	elements = append(elements, strings.TrimSpace(json[start:]))
	return elements
}
